pydantic_ai_slim/pydantic_ai/agent/__init__.py

```diff
@@ -1413,13 +1459,24 @@ def copy_dynamic_toolsets(toolset: AbstractToolset[AgentDepsT]) -> AbstractTools
 
         toolset = toolset.visit_and_replace(copy_dynamic_toolsets)
 
+        tool_config = (
+            effective_prompt_config.tool_config
+            if (effective_prompt_config := self._get_prompt_config(prompt_config))
+            else None
+        )
+
         if self._prepare_tools:
             toolset = PreparedToolset(toolset, self._prepare_tools)
 
+        if tool_config:
+            toolset = ToolConfigPreparedToolset(toolset, tool_config)
+
         output_toolset = output_toolset if _utils.is_set(output_toolset) else self._output_toolset
         if output_toolset is not None:
             if self._prepare_output_tools:
                 output_toolset = PreparedToolset(output_toolset, self._prepare_output_tools)
+            if tool_config:
+                output_toolset = ToolConfigPreparedToolset(output_toolset, tool_config)
```

Comment:
Can't we add this once at the very end, applying to the entire `toolset` just before we return it?